zebra-consensus/src/verify.rs

```diff
@@ -0,0 +1,76 @@
+//! Block verification and chain state updates for Zebra.
+//!
+//! Verification occurs in multiple stages:
+//!   - getting blocks (disk- or network-bound)
+//!   - context-free verification of signatures, proofs, and scripts (CPU-bound)
+//!   - context-dependent verification of the chain state (awaits a verified parent block)
+//!
+//! Verification is provided via a `tower::Service`, to support backpressure and batch
+//! verification.
+
+use std::{
+    error::Error,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+
+mod script;
+mod transaction;
+
+/// Block verification service.
+///
+/// After verification, blocks and their associated transactions are added to
+/// `zebra_state::ZebraState`.
+#[derive(Default)]
+struct BlockVerifier {}
+
+/// The result type for the BlockVerifier Service.
+type Response = BlockHeaderHash;
+
+/// The error type for the BlockVerifier Service.
+// TODO(jlusby): Error = Report ?
+type ServiceError = Box<dyn Error + Send + Sync + 'static>;
+
+/// The BlockVerifier service implementation.
+impl Service<Block> for BlockVerifier {
+    type Response = Response;
+    type Error = ServiceError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, context: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // TODO(teor): is this a shared state?
+        let mut state_service = zebra_state::in_memory::init();
```

Comment:
I think this one needs to be long lived, we should store the service that this returns in the `BlockVerifier` and then call `in_memory::init` inside of `BlockVerifier::init` when constructing the block verifier service.

Reply:
I'm struggling with the types and lifetimes in this change.

Commit 30b1c848046b0c499c2af60ee577173459039eba doesn't have this change, it compiles fine.

Commit d47e77b02c70ae9454d708159e0710b8ac60c141 has my attempt at this change, with `error: the `call` method cannot be invoked on a trait object`. The full error is in the commit message.

Reply:
CI also shows an extra error: "the `poll_ready` method cannot be invoked on a trait object"
https://github.com/ZcashFoundation/zebra/pull/448/checks?check_run_id=757541268#step:5:759

Reply:
Fixed in d19fd2ec504aa4fc813338ae4e0b5f1dbb169294.